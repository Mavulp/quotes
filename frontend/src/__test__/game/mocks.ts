import type { Player, RoundFillQuote, RoundGuessQuotee } from '../../types/game-types'

export const MOCK_ROUND_FILL: RoundFillQuote = {
  originalQuote: {
    id: 571,
    author: 'Jokler',
    offensive: true,
    indices: [
      {
        quotee: 'tmtu',
        index: 112,
      },
    ],
    fragments: [
      {
        type: 'text',
        highlight: true,
        content: '[That might not be what you said but] that is what I heard',
        quotee: 'tmtu',
      },
    ],
    tags: [
      'legacy',
    ],
    createdAt: 1563822564,
  },
  fragmentIndex: 0,
  time: 30,
  answers: [
    4,
    10,
  ],
  words: [
    '[That',
    'might',
    'not',
    'be',
    'what',
    'you',
    'said',
    'but]',
    'that',
    'is',
    'what',
    'I',
    'heard',
  ],
  difficulty: 'Medium',
  type: 'fill-the-quote',
}

export const MOCK_PLAYER: Player = {
  username: 'tester',
  score: 0,
  ready: false,
  _input: null,
  _inputTimestamp: -1,
}

export const MOCK_ROUND_QUOTEE: RoundGuessQuotee = {
  options: [
    'UNKNOWN',
    'tmtu',
    'Frank',
    'kilmanio',
  ],
  answer: 'tmtu',
  fragmentIndex: 0,
  originalQuote: {
    id: 1081,
    author: 'Jokler',
    offensive: true,
    indices: [
      {
        quotee: 'tmtu',
        index: 169,
      },
    ],
    fragments: [
      {
        type: 'text',
        highlight: true,
        content: '[Marina sounds extra slutty today] Truer words have never been spoken.',
        quotee: 'tmtu',
      },
    ],
    tags: [
      'legacy',
    ],
    createdAt: 1581619013,
  },
  time: 30,
  difficulty: 'Medium',
  type: 'guess-the-quotee',
}
