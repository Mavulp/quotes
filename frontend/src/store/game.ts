import { defineStore } from 'pinia'
import { computed, reactive, ref, unref } from 'vue'
import { isNil, orderBy, partition, result, shuffle } from 'lodash'
import type { Difficulty, Fragment, GameState, Gamemode, Player, RoundAnswer, RoundResults, RoundTypes } from '../types/game-types'
import { useQuote } from '../store/quote'
import type { ValueOf } from '../bin/utils'
import { arrayIntoChunks, getRanMinMax } from '../bin/utils'
import type { Quote } from '../types/quote-types'
import { useUser } from './user'

// TODO properly document
// TODO Add round history, per player
//          - save their answer (and if it was correct)
//          - how many points they goet
//          - how long it took to answer

// TODO [refactor] Instead of pasting in originalQuote, use the ID

// FIXME
// Exclude quotes which are less than 4 words long and or contain a url for fill-the-quote

// FIXME
// Filtering quotes down crashes

// TODO Split all defaults (mainly in reset functions) into separate exported variables or defaults.ts file
// TODO in endRound(); detect if it was the last round and perform game end func

// TODO support for validating multiple answers (validateAnswer should not
// return a simple boolean). Also save user inputs
// This implementation was started with RoundAnswer & RoundResults

export const difficultyOptions: Difficulty[] = ['Easy', 'Medium', 'Hard']
export const gamemodeOptions = [
  { value: 'guess-the-quotee', label: 'Guess The Quotee' },
  { value: 'guess-the-author', label: 'Guess The Author' },
  { value: 'fill-the-quote', label: 'Fill The Quote' },
]
const gamemodeIds = Object.values(gamemodeOptions).map(g => g.value) as Gamemode[]
const gamemodeAmount = Object.keys(gamemodeOptions).length

export const useGame = defineStore('game', () => {
  // Constants
  const BASE_POINTS = 100
  const TRANSITION_DELAY_S = 8

  // Manage history
  const state = reactive<GameState>({} as GameState)
  const players = ref<Player[]>([])
  const cfg = reactive({
    // Global Settings
    maxPlayerCount: 8,
    globalRoundLength: 30,
    difficulty: 'Medium' as Difficulty,
    // Quote pool settings
    useCustomPool: false,
    // Game composition settings
    useCustomComposition: false,
    // Thse settings are only used if customComposition is OFF
    rounds: 10,
  })
  // This contains the game
  const fragments = ref<Fragment[]>([])

  function resetState(gameId: string, admin: string) {
    Object.assign(state, {
      gameId,
      admin,
      paused: false,
      stage: 'setup',
      quotePool: new Set(),
      transformedPool: [],
      roundIndex: 0,
      startTime: -1,
      endTime: -1,
      history: [],
    })

    addPlayer(admin)
  }

  function resetConfig() {
    Object.assign(cfg, {
      maxPlayerCount: 8,
      globalRoundLength: 30,
      difficulty: 'Medium',
      useCustomPool: false,
      useCustomComposition: false,
      rounds: 10,
    })
  }

  function addPlayer(username: string) {
    players.value.push({
      username,
      score: 0,
      ready: false,
      _input: null,
      _inputTimestamp: -1,
    })
  }

  function resetPlayersAtRoundEnd() {
    for (const player of players.value) {
      player._input = null
      player._inputTimestamp = -1
      player.ready = false
    }
  }

  /**
   * History methods
   */

  function addHistoryEntry(round: RoundTypes, points: RoundResults[]) {
    const item = {
      ...round,
      points,
    }
    state.history.push(item)
  }

  function saveHistory() {
    // Collect relevant information from the game states into an object

    // Makes an API call at the end of the game, sending the whole game object
  }

  function setPlayerState(username: string, key: keyof Player, value: ValueOf<Player>) {
    const index = players.value.findIndex(p => p.username === username)

    // If input is being updated, set the input timestamp as well to track
    // when player has answered
    if (key === '_input')
      Reflect.set(players.value[index], '_inputTimestamp', Date.now())

    Reflect.set(players.value[index], key, value)
  }

  function insertFragment() {
    fragments.value.push({
      gamemode: 'guess-the-quotee',
      rounds: 8,
      roundTime: 30,
      difficulty: unref(cfg.difficulty) as Difficulty,
    })
  }

  function removeFragment(index: number) {
    fragments.value.splice(index, 1)
  }

  function createQuotePool() {
    const quotes = useQuote()
    //
    const quoteCount = cfg.useCustomPool
      ? fragments.value.reduce((count, fragment) => count += fragment.rounds, 0)
      : cfg.rounds

    // To make generating easier, just use set to store N amount of quotes
    const quoteIdPool: Set<number> = new Set()

    while (quoteIdPool.size < quoteCount) {
      const index = getRanMinMax(0, quotes.quotes.length)
      const quote = quotes.quotes[index]
      quoteIdPool.add(quote.id)
    }

    state.quotePool = quoteIdPool
    return quoteIdPool
  }

  // Takes in a quote pool and returns task pool
  function transformQuotes() {
    const quotes = useQuote()
    const transformed: RoundTypes[] = []

    if (cfg.useCustomComposition) {
      const pool = [...state.quotePool]

      // Iterate over fragments
      for (const fragment of fragments.value) {
        for (let i = 0; i < fragment.rounds; i++) {
          // Get first quote from the pool
          const quoteId = pool.shift()
          const quote = quotes.getQuoteById(quoteId)

          if (!quote)
            continue

          transformed.push(
            transformQuote(
              quote,
              fragment.gamemode,
              fragment.difficulty,
              fragment.roundTime,
            ),
          )
        }
      }
    }
    else {
      // Split amount of rounds by the amount of gamemodes
      const gamemdes = shuffle(gamemodeIds)

      const groups = arrayIntoChunks(
        [...state.quotePool],
        gamemodeAmount,
      )

      groups.forEach((group, index) => {
        for (const quoteId of group) {
          const quote = quotes.getQuoteById(quoteId)

          if (!quote)
            continue

          transformed.push(
            transformQuote(
              quote,
              gamemdes[index],
              cfg.difficulty,
              cfg.globalRoundLength,
            ),
          )
        }
      })
    }

    state.transformedPool = transformed

    return transformed
  }

  // Takes in a raw quote and returns a transformed Task
  function transformQuote(quote: Quote, type: Gamemode, difficulty: Difficulty, time: number): RoundTypes {
    // Select relevant quote fragment
    const selectedFragment = orderBy(quote.fragments, ['type', 'highlight'], ['desc', 'desc'])[0]
    const { content, quotee } = selectedFragment
    // Index of the selected fragment in the original quote
    const fragmentIndex = quote.fragments.findIndex(f => f.content === selectedFragment.content)

    // Based on difficulty
    // index of Easy is 0 (+1) deems 1 iteration and so on
    let len = difficultyOptions.indexOf(difficulty) + 1

    switch (type) {
      case 'fill-the-quote': {
        // Take the quote and based on difficulty extract 1-3 words from it
        const answers: Set<number> = new Set()
        const words = content.trim().split(/\s+/)

        while (answers.size !== len)
          answers.add(getRanMinMax(0, words.length - 1))

        return {
          originalQuote: quote,
          fragmentIndex,
          time,
          answers: [...answers],
          words,
          difficulty,
          type: 'fill-the-quote',
        }
      }
      case 'guess-the-author': {
        const users = useUser()
        const answer = quote.author

        // Slightly increase difficulty when you only have to choose a user
        len *= 2

        const options = new Set<string>()
        const cloned = [...users.users].filter(user => user.username !== answer)

        // Add the answer
        options.add(answer)

        while (options.size !== len)
          options.add(cloned[getRanMinMax(0, cloned.length - 1)].username)

        return {
          answer,
          options: shuffle([...options]),
          originalQuote: quote,
          time,
          difficulty,
          type: 'guess-the-author',
        }
      }

      case 'guess-the-quotee': {
        const users = useUser()

        // Slightly increase difficulty when you only have to choose a user
        len *= 2

        const options = new Set<string>()
        const cloned = [...users.users].filter(user => user.username !== quotee)

        // Add the answer
        options.add(quotee)

        while (options.size !== len)
          options.add(cloned[getRanMinMax(0, cloned.length - 1)].username)

        return {
          options: shuffle([...options]),
          answer: quotee,
          fragmentIndex,
          originalQuote: quote,
          time,
          difficulty,
          type: 'guess-the-quotee',
        }
      }
    }
  }

  // Iterates over players and checks their input against
  // REVIEW Should validation also store which answers players got right and wrong? (with attached inputs)
  function validatePlayerAnswers(players: Player[], round: RoundTypes) {
    // const round = state.transformedPool[state.roundIndex]
    const formattedResults = players
      // 1. Iterate over players and check answers as correct or incorrect
      .map(p => ({
        results: validateAnswer(p, round),
        player: p,
      }))

    const [correctResults, wrongResults] = partition(formattedResults, r => r.results.some(res => res.correct))

    // The return order matters because it already determines the order of
    // players in the post-round screen
    return [
      // ----- CORRECT results
      ...correctResults
        .sort((a, b) => a.player._inputTimestamp > b.player._inputTimestamp ? -1 : 1)
        .map((result, index) => {
          // Calculate score based on results
          const totalAnswers = result.results.length
          const correctAnswers = result.results.filter(r => r.correct).length
          // Additional points
          // TODO: make better
          const bonusScore = BASE_POINTS * (correctAnswers / totalAnswers)

          const score = (index + 1) * BASE_POINTS + bonusScore

          const scoreBefore = result.player.score
          result.player.score += score

          return {
            scoreBefore,
            score,
            username: result.player.username,
          }
        }),
      // ----- WRONG results
      ...wrongResults.map(result => ({
        scoreBefore: result.player.score,
        score: result.player.score,
        username: result.player.username,
      })),
    ]
  }

  // Compare player input to the round answers
  function validateAnswer(player: Player, round: RoundTypes): RoundAnswer[] {
    if (round.type === 'fill-the-quote') {
      const input = player._input as Record<number, string> | null

      if (!input)
        return [{ input: null, correct: false }]

      return Object.entries(input).map(([index, value]) => {
        return {
          input: value,
          correct: value ? round.words[Number(index)].toLowerCase() === value.toLowerCase() : false,
        }
      })
    }

    return [{
      input: player._input,
      correct: player?._input?.toLowerCase() === round.answer.toLowerCase() ?? false,
    }]
  }

  /**
   * Computed properties
   */

  // This is used to check if players are ready to start the game
  // Same property can also be used in game, if everyone answers before time runs out
  // and clicks the "final answer" button, it will lock in their answer and set them as ready
  const arePlayersReady = computed(() => players.value.every(p => p.ready))

  function setPlayersNotReady() {
    for (const player of players.value)
      player.ready = false
  }

  return {
    cfg,
    state,
    players,
    fragments,
    addPlayer,
    resetState,
    resetConfig,
    saveHistory,
    validateAnswer,
    insertFragment,
    removeFragment,
    setPlayerState,
    addHistoryEntry,
    transformQuotes,
    createQuotePool,
    arePlayersReady,
    setPlayersNotReady,
    validatePlayerAnswers,
    resetPlayersAtRoundEnd,

    // Constants
    BASE_POINTS,
    TRANSITION_DELAY_S,
  }
})
