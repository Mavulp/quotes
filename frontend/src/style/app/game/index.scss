@import "./configure";
@import "./ingame";
@import "./fragments";

.route-game {
  padding-top: 20px;
}
