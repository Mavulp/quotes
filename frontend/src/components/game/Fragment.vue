<script setup lang='ts'>
import { computed } from 'vue'
import { difficultyOptions, gamemodeOptions, useGame } from '../../store/game'
import type { Fragment } from '../../types/game-types'
import InputSelect from '../form/InputSelect.vue'
import InputText from '../form/InputText.vue'

const props = defineProps<{
  fragment: Fragment
  index: number
}>()
const game = useGame()

function editKey(key: keyof Fragment) {
  return computed({
    get: () => {
      const f = game.fragments[props.index]
      return Reflect.get(f, key) as string
    },
    set: (value) => {
      const f = game.fragments[props.index]
      Reflect.set(f, key, value)
    },
  })
}

const type = editKey('type')
const rounds = editKey('rounds')
const roundTime = editKey('roundTime')
const difficulty = editKey('difficulty')
</script>

<template>
  <div class="composition-fragment">
    <span class="fragment-number">#{{ props.index + 1 }}</span>

    <button class="remove-fragment button btn-round btn-white" @click="game.removeFragment(props.index)">
      <Icon code="e5cd" size="2" />
    </button>

    <div class="content-cell">
      <label>Gamemode</label>
      <InputSelect v-model:selected="type" :options="gamemodeOptions" />
    </div>

    <div class="content-cell">
      <label>Repeats</label>
      <InputText v-model:value="rounds" type="number" min="1" />
    </div>

    <div class="content-cell">
      <label>Round time limit (seconds)</label>
      <InputText v-model:value="roundTime" type="number" min="5" />
    </div>

    <div class="content-cell">
      <label>Difficulty</label>
      <InputSelect v-model:selected="difficulty" :options="difficultyOptions" />
    </div>
  </div>
</template>
