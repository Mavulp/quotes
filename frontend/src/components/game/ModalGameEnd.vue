<script setup lang='ts'>
import { useGame } from '../../store/game'

const game = useGame()
</script>

<template>
  <div class="quote-container-small game-end-modal">
    <div class="modal-content">
      <h3>That's a wrap!</h3>

      <pre>
        {{ game.state }}
      </pre>
    </div>
  </div>
</template>
