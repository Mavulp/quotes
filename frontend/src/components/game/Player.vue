<script setup lang='ts'>
import { computed, ref } from 'vue'
import { getRndGradient } from '../../bin/color'
import { useGame } from '../../store/game'
import { useUser } from '../../store/user'

const props = defineProps<{ username: string }>()
const users = useUser()
const game = useGame()

const user = computed(() => users.users.find(u => u.username === props.username))
const isAdmin = computed(() => props.username === game.state.admin)
</script>

<template>
  <div v-if="user" class="game-user" :class="{ 'is-admin': isAdmin }">
    <div class="image-wrap">
      <img v-if="user.profilePicture" :src="user.profilePicture">
      <div v-else :style="{ backgroundColor: getRndGradient() }">
        {{ user.username[0] }}
      </div>
    </div>

    <span>
      {{ user.username }}
      &nbsp;
      <div v-if="isAdmin" class="tag highlight">Admin</div>
    </span>

    <!-- Display admin controls on all users only for admin & and not on admin themselves -->
    <div v-if="users.username === game.state.admin && props.username !== users.username" class="user-attrs">
      <button class="button btn-gray btn-round" data-title-top="Kick user" @click="game.removePlayer(props.username)">
        <Icon code="ef66" size="2" />
      </button>

      <button class="button btn-gray btn-round" data-title-top="Make Admin" @click="game.state.admin = props.username">
        <Icon code="ea3d" size="2" />
      </button>
    </div>
  </div>
</template>
