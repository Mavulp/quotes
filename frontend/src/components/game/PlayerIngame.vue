<script setup lang='ts'>
import { isNil } from 'lodash'
import { computed } from 'vue'
import { getRndGradient } from '../../bin/color'
import { useUser } from '../../store/user'
import type { Player } from '../../types/game-types'

const props = defineProps<{
  player: Player
  difference?: number
}>()

const users = useUser()
const user = computed(() => users.users.find(u => u.username === props.player.username))
</script>

<template>
  <div v-if="user" class="game-user">
    <div class="image-wrap">
      <img v-if="user.profilePicture" :src="user.profilePicture">
      <div v-else :style="{ backgroundColor: getRndGradient() }">
        {{ user.username[0] }}
      </div>
    </div>

    <div class="game-context">
      <span>{{ user.username }}</span>
      <slot />
    </div>

    <strong>
      {{ props.player.score }}
    </strong>
    <span
      v-if="!isNil(props.difference)"
      class="tag small"
      :class="[props.difference === 0 ? 'red' : 'green']"
    >
      +{{ props.difference }}
    </span>
  </div>
</template>
