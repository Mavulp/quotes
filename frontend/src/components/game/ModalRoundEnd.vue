<script setup lang='ts'>
import Countdown from '@chenfengyuan/vue-countdown'
import dayjs from 'dayjs'
import { Player } from '../../types/game-types'
import type { RoundResults } from '../../types/game-types'
import { useGame } from '../../store/game'
import PlayerIngame from './PlayerIngame.vue'

const props = defineProps<{ results: RoundResults[] }>()
const game = useGame()
const round = game.state.transformedPool[game.state.roundIndex]
const timer = dayjs.utc().add(game.TRANSITION_DELAY_S - 1 + game.players.length, 'second').diff(Date.now())
</script>

<template>
  <div class="quote-container-small game-transition-modal">
    <div class="modal-content">
      <h3>Round {{ game.state.roundIndex + 1 }}</h3>

      <div class="game-modal-question">
        <!-- Fill the quote -->
        <template v-if="round.type === 'fill-the-quote'">
          <span>What {{ round.originalQuote.fragments[round.fragmentIndex].quotee }} actually said...</span>
          <div class="quote-fill-wrap">
            <template v-for="(word, index) in round.words" :key="word">
              <b v-if="round.answers.includes(index)" v-text="word" />
              <span v-else>{{ word }}</span>
            </template>
          </div>
        </template>

        <!-- Guess the author -->
        <template v-else-if="round.type === 'guess-the-author'">
          <div v-for="fragment in round.originalQuote.fragments" :key="fragment.content" class="fragment-author">
            <p>{{ fragment.content }}</p>
            <span class="fragment-quotee">
              <Icon code="e244" size="1.4" />
              {{ fragment.quotee }}
            </span>
          </div>
          <span>The author is...</span>
          <strong>{{ round.originalQuote.author }}</strong>
        </template>

        <!-- Guess the quotee -->
        <template v-else>
          <p>{{ round.originalQuote.fragments[round.fragmentIndex].content }}</p>
          <span>The one who said it is...</span>
          <strong> {{ round.originalQuote.fragments[round.fragmentIndex].quotee }}</strong>
        </template>
      </div>

      <div class="player-list">
        <div v-for="(item, index) in props.results" :key="item.username" class="game-user-wrap">
          <div class="game-user-wrap-inner">
            <span class="rank">#{{ index + 1 }}</span>
            <PlayerIngame
              class="in-modal"
              :player="{
                score: item.score,
                username: item.username,
              // Considering PlayerInagme only uses score and username
              //we can type cast it here
              } as Player"
              :difference="item.score === item.scoreBefore ? 0 : item.score - item.scoreBefore"
            >
              <div class="game-user-answer">
                <span v-if="item.answers.every(a => !a.input)">No answer</span>
                <template v-else>
                  <span>Answered:</span>
                  <span v-for="answer in item.answers" :key="answer.input" :class="[answer.correct ? 'green' : 'red']">
                    {{ answer.input }}
                  </span>
                </template>
              </div>
            </PlayerIngame>
          </div>
        </div>
      </div>
      <hr>
      <p>
        Next round in
        <Countdown v-slot="{ seconds }" :time="timer">
          {{ seconds }}
        </Countdown> seconds
      </p>
    </div>
  </div>
</template>
