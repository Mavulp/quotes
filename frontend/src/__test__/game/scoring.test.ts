import { createPinia, setActivePinia } from 'pinia'
import { beforeEach, describe, expect, test } from 'vitest'
import { useGame } from '../../store/game'
import { MOCK_PLAYER, MOCK_ROUND_FILL, MOCK_ROUND_QUOTEE } from './mocks'

describe('Iterate over provided answers and assign poinst to plays', () => {
  beforeEach(() => {
    setActivePinia(createPinia())
  })

  test('Validate simple answer', () => {
    const game = useGame()
    const player = {
      ...MOCK_PLAYER,
      _input: 'tmtu',
    }
    const result = game.validateAnswer(player, MOCK_ROUND_QUOTEE)
    expect(result).toStrictEqual([{
      input: 'tmtu',
      correct: true,
    }])
  })

  test('Validate complex answer', () => {
    const game = useGame()
    const player = {
      ...MOCK_PLAYER,
      _input: {
        4: 'what',
        10: 'that',
      },
    }
    const result = game.validateAnswer(player, MOCK_ROUND_FILL)
    expect(result).toStrictEqual([
      {
        input: 'what',
        correct: true,
      },
      {
        input: 'that',
        correct: false,
      },
    ])
  })

  test('Validate player answers', () => {
    const game = useGame()

    const players = [
      {
        // Answered second, is right
        ...MOCK_PLAYER,
        username: 'Anna',
        _input: 'tmtu',
        _inputTimestamp: Date.now() - 100,
        score: 50,
      },
      // Answered last, is right
      {
        ...MOCK_PLAYER,
        username: 'Jeniffer',
        _input: 'tmtu',
        _inputTimestamp: Date.now(),
        score: 0,
      },
      // Answered first, is wrong
      {
        ...MOCK_PLAYER,
        username: 'Pamela',
        _input: 'Franku',
        _inputTimestamp: Date.now() - 300,
        score: 200,
      },
      // Answered last, is wrong
      {
        ...MOCK_PLAYER,
        username: 'Daniel',
        _input: null,
        _inputTimestamp: Date.now(),
        score: 100,
      },
    ]

    const results = game.validatePlayerAnswers(players, MOCK_ROUND_QUOTEE)
    expect(results).toMatchSnapshot()
  })
})
