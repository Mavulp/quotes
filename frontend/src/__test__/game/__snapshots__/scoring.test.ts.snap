// Vitest Snapshot v1

exports[`Iterate over provided answers and assign poinst to plays > Validate player answers 1`] = `
[
  {
    "score": 100,
    "scoreBefore": 0,
    "username": "Jeniffer",
  },
  {
    "score": 200,
    "scoreBefore": 50,
    "username": "Anna",
  },
  {
    "score": 200,
    "scoreBefore": 200,
    "username": "Pamela",
  },
  {
    "score": 100,
    "scoreBefore": 100,
    "username": "Daniel",
  },
]
`;
